sw/airborne/modules/decawave/decawave_serial_communication.c

```diff
@@ -0,0 +1,268 @@
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/decawave_serial_communication.h"
+ * @author S. vd H, C. DW
+ *
+  Decawave DWM1000 module serial communication for use in anchorless network where the UWB modules are attached to MAVs and need to communicate 
+  on-board values with each-other (for purposes such as relative localization, co-ordination, or collision avoidance).
+  This module must be used together with the Decawave DWM1000 running the appropriate Serial Communication code, which can be flashed on the Arduino board.
+  The arduino library can be found at:
+    https://github.com/StevenH2812/arduino-dw1000/tree/UWB_localization_v1.0
+  The example file to flash the Arduino Micro can be found in
+    examples/UWB_localization_v1_0/UWB_localization_v1_0.ino
+ */
+
+#include "decawave_serial_communication.h"
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/radio_control.h"
+#include "state.h"
+#include "pprzlink/messages.h"
+#include "subsystems/datalink/downlink.h"
+#include "mcu_periph/uart.h"
+#include <stdio.h>
+#include "subsystems/abi.h"
+
+PRINT_CONFIG_VAR(SERIAL_UART)
+PRINT_CONFIG_VAR(SERIAL_BAUD)
+
+// define coms link for stereocam
+#define SERIAL_PORT   (&((SERIAL_UART).device))
+struct link_device *xdev = SERIAL_PORT;
+
+#define SerialGetch() SERIAL_PORT ->get_byte(SERIAL_PORT->periph)
+#define SerialSend1(c) SERIAL_PORT->put_byte(SERIAL_PORT->periph, 0, c)
+#define SerialUartSend1(c) SerialSend1(c)
+#define SerialSend(_dat,_len) { for (uint8_t i = 0; i< (_len); i++) SerialSend1(_dat[i]); };
+#define SerialUartSetBaudrate(_b) uart_periph_set_baudrate(SERIAL_PORT, _b);
+#define SerialChAvailable()(xdev->char_available(xdev->periph))
+#define SerialSendNow() uart_send_message(SERIAL_PORT->periph,0)
+
+struct nodeState{
+  uint8_t nodeAddress;
+  float vx;
+  float vy;
+  float z;
+  float r;
+  bool stateUpdated[UWB_SERIAL_COMM_NODE_STATE_SIZE];
+};
+
+static uint8_t _bytesRecvd = 0;
+static uint8_t _dataRecvCount = 0;
+
+static uint8_t _tempBuffer[UWB_SERIAL_COMM_MAX_MESSAGE];
+static uint8_t _tempBuffer2[UWB_SERIAL_COMM_MAX_MESSAGE];
+static uint8_t _recvBuffer[UWB_SERIAL_COMM_FLOAT_SIZE];
+
+static uint8_t _dataSendCount = 0;
+static uint8_t _dataTotalSend = 0;
+
+static bool _inProgress = false;
+static bool _allReceived = false;
+
+static uint8_t _varByte = 0;
+
+static struct nodeState _states[UWB_SERIAL_COMM_DIST_NUM_NODES];
+float range_float = 0.0;
+
```

Comment:
In general, you are using here a lot of global variables that could be gathered in a single structure with the buffers, status flags, indexes, uart device pointer, ...